crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -8140,7 +8140,36 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                     slice_ty,
                 )
             }
+            // If the value type is a union make sure to union the load types.
+            // For example:
+            // val: tuple[int] | tuple[str]
+            // val[0] can be an int or str type
+            (Type::Union(union_ty), _, _) => {
+                let value_types = union_ty.elements(self.db());
+                let loads_types: Vec<Type<'db>> = value_types
+                    .iter()
+                    .copied()
+                    .map(|ty| self.infer_subscript_expression_types(value_node, ty, slice_ty))
+                    .collect();
```

Comment:
You don't need the `collect` call here, because `UnionType::from_elements` can take in the iterator directly

Reply:
This actually turns out to be a little harder than that, due to `self` being captured by the closure. But it's easily fixed by making `infer_subscript_expression_types` (and one other method that it calls) take `&self` instead of `&mut self` -- they don't actually need to store anything in the `TypeInferenceBuilder`.